cdap-master/src/main/java/co/cask/cdap/data/runtime/main/StartupChecker.java

```diff
@@ -0,0 +1,179 @@
+import org.apache.hadoop.yarn.client.api.YarnClient;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Performs some sanity checks to catch any glaring problems that would prevent master from starting up.
+ */
+public class StartupChecker extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(StartupChecker.class);
+  private final Configuration hConf;
+  private final CConfiguration cConf;
+  private final LocationFactory locationFactory;
+  // used because some operations hang if the dependent service is not running.
+  private final ExecutorService executorService;
+
+  @Inject
+  StartupChecker(Configuration hConf, CConfiguration cConf, LocationFactory locationFactory) {
+    this.hConf = hConf;
+    this.cConf = cConf;
+    this.locationFactory = locationFactory;
+    this.executorService = Executors.newSingleThreadExecutor(
+      new ThreadFactoryBuilder().setNameFormat("startup-checker").build());
+  }
+
+  /**
+   * Run sanity checks to see if there are any glaring problems that will prevent master from starting up.
+   * These are the types of things that are not likely to be resolved without manual intervention, such as incorrect
+   * directory permissions, wrong hive dependencies, and underlying infrastructure like HDFS being unavailable.
+   *
+   * Other types of errors that may eventually resolve themselves, such as failure to connect to transaction service,
+   * are handled in a separate method.
+   *
+   * This method is a no-op if {@link Constants#STARTUP_CHECKS_ENABLED} is false.
+   *
+   * @return whether the sanity checks succeeded or not
+   */
+  public boolean runSanityChecks() {
+    boolean noErrors = true;
+
+    ConfigurationLogger.logImportantConfig(cConf);
+    if (cConf.getBoolean(Constants.STARTUP_CHECKS_ENABLED)) {
+      noErrors = checkHDFS();
+      noErrors = checkYarn() && noErrors;
+      noErrors = checkHBase() && noErrors;
+    }
+
+    return noErrors;
+  }
+
+  private boolean checkHDFS() {
+    LOG.info("Checking that HDFS is available.");
+    FileSystem hdfs;
+    try {
+      hdfs = FileSystem.get(hConf);
+    } catch (Exception e) {
+      LOG.error("Unable to get an HDFS FileSystem instance. " +
+                  "Please check that the correct Hadoop libraries and configuration are included in the classpath.", e);
+      return false;
+    }
+
+    try {
+      hdfs.getStatus();
+    } catch (Exception e) {
+      LOG.error("Unable to get the status of HDFS. Please check that HDFS is running, " +
+                  "and that the correct Hadoop libraries and configuration are included in the classpath.", e);
```

Comment:
Here you don't say CDAP Master classpath, but all messages below say that. Should we be consistent ?